tikvrpc/interceptor/interceptor.go

```diff
@@ -0,0 +1,185 @@
+//     }
+// }
+// txn.SetRPCInterceptor(LogInterceptor)
+// ```
+//
+// Or you want to inject some dependent modules:
+// ```
+// func GetLogInterceptor(lg *log.Logger) RPCInterceptor {
+//     return func(next RPCInterceptorFunc) RPCInterceptorFunc {
+//         return func(target string, req *tikvrpc.Request) (*tikvrpc.Response, error) {
+//             lg.Println("before")
+//             resp, err := next(target, req)
+//             lg.Println("after")
+//             return resp, err
+//         }
+//     }
+// }
+// txn.SetRPCInterceptor(GetLogInterceptor())
+// ```
+type RPCInterceptor func(next RPCInterceptorFunc) RPCInterceptorFunc
+
+// RPCInterceptorFunc is a callable function used to initiate a request to TiKV.
+// It is mainly used as the parameter and return value of RPCInterceptor.
+type RPCInterceptorFunc func(target string, req *tikvrpc.Request) (*tikvrpc.Response, error)
+
+// RPCInterceptorChain is used to combine multiple interceptors into one.
+// Multiple interceptors will be executed in the order of link time, but are more
+// similar to the onion model: The earlier the interceptor is executed, the later
+// it will return.
+//
+// We can use RPCInterceptorChain like this:
+// ```
+// func Interceptor1(next InterceptorFunc) RPCInterceptorFunc {
+//     return func(target string, req *tikvrpc.Request) (*tikvrpc.Response, error) {
+//         fmt.Println("begin-interceptor-1")
+//         defer fmt.Println("end-interceptor-1")
+//         return next(target, req)
+//     }
+// }
+// func Interceptor2(next InterceptorFunc) RPCInterceptorFunc {
+//     return func(target string, req *tikvrpc.Request) (*tikvrpc.Response, error) {
+//         fmt.Println("begin-interceptor-2")
+//         defer fmt.Println("end-interceptor-2")
+//         return next(target, req)
+//     }
+// }
+// txn.SetRPCInterceptor(NewRPCInterceptorChain().Link(Interceptor1).Link(Interceptor2).Build())
+// ```
+//
+// Then every time an RPC request is initiated, the following text will be printed:
+// ```
+// begin-interceptor-1
+// begin-interceptor-2
+// /* do request & respond here */
+// end-interceptor-2
+// end-interceptor-1
+// ```
+type RPCInterceptorChain struct {
+	chain []RPCInterceptor
+}
+
+// NewRPCInterceptorChain creates an empty RPCInterceptorChain.
+func NewRPCInterceptorChain() *RPCInterceptorChain {
+	return &RPCInterceptorChain{}
+}
+
+// Link is used to link the next RPCInterceptor.
+// Multiple interceptors will be executed in the order of link time.
+func (c *RPCInterceptorChain) Link(it RPCInterceptor) *RPCInterceptorChain {
+	c.chain = append(c.chain, it)
+	return c
+}
+
+// Build merges the previously linked interceptors into one.
+func (c *RPCInterceptorChain) Build() RPCInterceptor {
+	return func(next RPCInterceptorFunc) RPCInterceptorFunc {
+		for n := len(c.chain) - 1; n >= 0; n-- {
+			next = c.chain[n](next)
+		}
+		return next
+	}
+}
+
+// ChainRPCInterceptors chains multiple RPCInterceptors into one.
```

Comment:
Better to comment the order of calling these RPCInterceptors.

Reply:
done